adding outline of Image testing standards (#1440)
chainguard-dev/edu: content/chainguard/chainguard-images/images-testing.md

---
title: "How Chainguard Images are Tested"
linktitle: "How Images are Tested"
type: "article"
description: "A conceptual article outlining testing requirements for Chainguard Images."
date: 2024-03-13T11:07:52+02:00
lastmod: 2023-03-13T11:07:52+02:00
draft: false
tags: ["Overview", "Chainguard Images", "Product"]
images: []
menu:
  docs:
    parent: "chainguard-images"
weight: 550
toc: true
---

> Chainguard Images are minimal, distroless container images that you can use to build and run secure applications. A frequently-asked question about these Images relates to what kind of testing Chainguard performs on them to ensure that they are indeed smaller and more secure than external counterparts.

Reviewer: ```suggestion
Chainguard Images are minimal, distroless container images that you can use to build and run secure applications. A frequently-asked question about these Images relates to what kind of testing Chainguard performs on them to ensure they match the functionality of their upstream counterparts.
```

This article provides a high level overview of what testing Chainguard performs when building new Images to ensure their security and that they provide an experience consistent with their external counterparts.

## Build requirements for new Images

Chainguard has a set of requirements in place that new Images must meet in order to be included in our library of Images. These requirements fall into two categories:

* Image standards
* Application testing

### Image standards

When building a new Image, Chainguard will take steps to ensure it meets the following standards:

| **Requirement** 	  |  **Explanation**     |
| --- | --- |
| **Size**     |  Any new Chainguard Images should be smaller than their external counterparts. Any Images that aren't smaller than their counterparts must include an explanation as to why.     |
|  **CVEs**     | When scanned with a CVE scanning tool like Grype, new Images should return zero CVEs. Again, if an Image does return CVEs in its scan results it should include an explanation. 	  |

> |  **Kubernetes accessibility**     | Containers based on Images must be able to run inside of a Kubernetes cluster.     |

Reviewer: ```suggestion
|  **Kubernetes accessibility**     | Containers used for Kubernetes based deployments must be able to run inside of a Kubernetes cluster.     |
```

|  **Architecture**     | Chainguard Images must be built for both the `x86_64` and `aarch64` architectures.      |

> |  **Tags**     | Chainguard Image tags must be configured correctly and align with their respective versions.      |

Reviewer: ```suggestion
```

I would remove this, we don't really make any commitments to aligning our tags with upstream(s), and have been bitten a few times for implying this. many times it works out, but sometimes it doens't

### Application testing

Chainguard performs the following checks on new Images to ensure that the applications contained within them meet the needs of most use cases:

| **Requirement** 	  |  **Explanation**     |
| --- | --- |

> | **Application startup**     |  When a container based on a Chainguard Image is launched, the Image's application starts correctly.    |

Reviewer: I would combine `Application startup`, `Functionality`, and `Helm chart compatbility` into just `Functionality`.

Since functionality varies based on the image, I think it makes more sense to bin it there, maybe something like:

```
|  **Functionality**     | The application is tested to comply with it's upstream counter parts core feature set.   |
```

> |  **Accessible**     | The Image's application is accessible to the user or the host system after startup.  |

Reviewer: ```suggestion
```

I would remove this, I think it's implied by the `Functionality` section below

|  **Functionality**     | The application's basic functionality is tested and documented.   |
|  **Helm chart compatibility**     | If there is an upstream Helm chart — or one provided by a customer — it will be tested to confirm that it can be used to install the application.      |

> |  **Builder Images**     | Chainguard's builder Images can in fact build new, functional images.     |

Reviewer: ```suggestion
|  **Toolchain Images**     | Images containing toolchains (`go`, `python`, etc...) can be used to build applications similar to their upstream counterparts.     |
```

## Automated tests

> In addition to the Image build requirements outlined previously, Chainguard also performs a number of automatic checks for new Images as part of our CI/CD process. These CI/CD checks ensure that every new Image can be installed with a provided Helm chart. As mentioned in the previous table, the Helm charts used for testing generally either come from a third-party source or a customer.

Reviewer: ```suggestion
In addition to the Image build requirements outlined previously, Chainguard also performs a number of automatic checks for new Images as part of our CI/CD process. 

Depending on the image, various representative (~functional/integration) tests are performed. For example, for applications primarily deployed via a Helm chart, the image is deployed to an ephemeral k8s cluster using the accepted helm chart and validated to perform in various ways.
```
